### PR TITLE
CXP-286: audit connections errors over a week

### PR DESCRIPTION
**Pull request by pchasle**
Add an internal API call to fetch audit data for the **Connections errors** chart
- Only for **auditable** Connections with a **data_source** flow type
- Number of errors per day for the **user timezone**
- Data is spread over **a week** (7 days) and include the **last day of the previous week** (+1 day)
- Data include the **total** number of error **for the current week**

API response look like this:
```json
{
  "erp": {
    "previous_week": {
      "2019-12-31": 3
    },
    "current_week": {
      "2020-01-01": 1,
      ...
      "2020-01-07": 7,
    },
    "current_week_total": 28
  }
}
```

### CODE REVIEW

File: src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+use Akeneo\Connectivity\Connection\Domain\ValueObject\DateTimePeriod;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+interface SelectPeriodErrorCountPerConnectionQuery
+{
+    /**
+     * @return PeriodEventCount[]
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php:17]:
```suggestion
     * @return PeriodErrorCount[]
```

File: src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Akeneo\Connectivity\Connection\Domain\Audit\Persistence\Query;
+
+use Akeneo\Connectivity\Connection\Domain\Audit\Model\Read\PeriodEventCount;
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php:7]:
```suggestion
use Akeneo\Connectivity\Connection\Domain\Audit\Model\Read\PeriodErrorCount;
```

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php [hunk 3]
```diff
@@ -12,19 +12,21 @@
  * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
  * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
  */
-final class AggregateProductEventCounts
+final class AggregateAuditData
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php:15]: Maybe this class name is a little bit too "open"

File: src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php (hunk 1 above)
Comment (pchasle) [reply src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php:17]: PeriodErrorCount doesn't exist, I'm using the PeriodEventCount as the ReadModel is **exactly** the same for each of our Audit.
We may want to rename it though `PeriodSomethingCount` ?
In the meantime you could consider that an error is still an event from an Audit point of view...

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php [hunk 4]
```diff
@@ -24,19 +27,24 @@ class AuditController
     private $userContext;
 
     /** @var GetPeriodEventCountPerConnectionHandler */
-    private $countDailyEventsByConnectionHandler;
+    private $getPeriodEventCountPerConnectionHandler;
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php:30]: Oops I missed these ones during the name refactoring

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php (hunk 3 above)
Comment (pchasle) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php:15]: Yes, do you have any suggestion ? Maybe `AggregateWeeklyAudit`, but it doesn't change much.

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php [hunk 5]
```diff
@@ -81,8 +130,11 @@ public function getErrorCountPerConnection(Request $request): JsonResponse
         return new JsonResponse($data);
     }
 
-    private function createUserDateTimeInterval(string $endDateUser, \DateTimeZone $timezone, \DateInterval $dateInterval): array
-    {
+    private function createUserDateTimeInterval(
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php:133]: I know it's not a new function but I think we should move this code in a service (currently we can't test this function at all :/ )

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectPeriodErrorCountPerConnectionQuery.php [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\Types\Types;
+
+/**
+ * @copyright 2020 Akeneo SAS (http://www.akeneo.com)
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class DbalSelectPeriodErrorCountPerConnectionQuery implements SelectPeriodErrorCountPerConnectionQuery
+{
+    use PeriodEventCountTrait;
```
Comment (pierrejolly) [inline src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectPeriodErrorCountPerConnectionQuery.php:22]: Maybe if it's used by error and event it's more an `PeriodCountTrait` ?

File: src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php (hunk 1 above)
Comment (willy-ahva) [reply src/Akeneo/Connectivity/Connection/back/Domain/Audit/Persistence/Query/SelectPeriodErrorCountPerConnectionQuery.php:17]: What I did when I was in the ErrorManagement domain is to copy pasted classes, changing the name. It's useful for type hinting. In the audit domain, I think I prefer renaming it as we keep the same object in the same domain.

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php (hunk 3 above)
Comment (willy-ahva) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Audit/AggregateAuditData.php:15]: AggregatePeriodMachinCount ?

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php (hunk 5 above)
Comment (pchasle) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/InternalApi/Controller/AuditController.php:133]: I agree with you on this. But didn't want to add that to this PR

File: src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectPeriodErrorCountPerConnectionQuery.php (hunk 6 above)
Comment (pchasle) [reply src/Akeneo/Connectivity/Connection/back/Infrastructure/Persistence/Dbal/Query/DbalSelectPeriodErrorCountPerConnectionQuery.php:22]: Yes, we need to think on a new name, same as the ReadModel

**Pull request merged**